src/main/java/community/solace/spring/boot/starter/solaceclientconfig/KeyStoreFactory.java

```diff
@@ -34,14 +37,33 @@ KeyStore createClientKeyStore(final String privateKeyPem, final String clientCer
             if (certificates == null) {
                 return null;
             }
+
+            if (!privateKeyMatchesCertificate((RSAPrivateCrtKey) privateKey, certificates)) {
+                LOG.error("Non of the given certificates in SSL_CLIENT_CERT matches the give key SSL_PRIVATE_KEY");
+                return null;
+            }
+
             keyStore.setKeyEntry("pk", privateKey, INTERNAL_PASSWORD.toCharArray(), certificates);
+
             return keyStore;
         } catch (Exception exception) {
             LOG.warn("Error during keystore creation", exception);
             return null;
         }
     }
 
+    private boolean privateKeyMatchesCertificate(RSAPrivateCrtKey privateKey, Certificate[] certificates) {
+        for (Certificate certificate : certificates) {
+            RSAPublicKey publicKey = (RSAPublicKey) certificate.getPublicKey();
```

Comment:
The check works for RSA keys only. Java supports DSA, RSA and EC. Please find a generic test tool (with embedded tests) here: https://gist.github.com/marchof/4e62893b636837411994f6fcf6fc371c